### PR TITLE
feat: created generic Codepen Snippet component

### PR DESCRIPTION
**Pull request by ErickPetru**
## Description of Problem
Right now, more than 20 files along the documentation have Codepen embeds with small changes on each one, like this:

``` html
<p class="codepen" data-height="300" data-theme-id="39028" data-default-tab="result" data-user="Vue" data-slug-hash="KKpRVvJ" data-preview="true" data-editable="true" style="height: 300px; box-sizing: border-box; display: flex; align-items: center; justify-content: center; border: 2px solid; margin: 1em 0; padding: 1em;" data-pen-title="Attribute dynamic binding">
  <span>See the Pen <a href="https://codepen.io/team/Vue/pen/KKpRVvJ">
  Attribute dynamic binding</a> by Vue (<a href="https://codepen.io/Vue">@Vue</a>)
  on <a href="https://codepen.io">CodePen</a>.</span>
</p>
<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
```

This large block of copied-and-pasted content takes up a lot of space in each place and is hard to maintain.

Also, as a documentation translation maintainer, I feel how it's annoying to keep all embeds translated in the same way when we have different collaborators translating different files.

And last but not least, using Single File Components inside Markdown with VuePress is so cool! Why not take advantage of this tool one more time to improve this scenario?

## Proposed Solution
A generic Codepen Snippet component, with props covering all different situations where Codepen embeds are use a little different across the documentation, but with good defaults that cover the majority of the use cases, to keep the usage simple.

## Additional Information
I took the liberty of changing all the Codepen embeds currently used in the documentation to use the new component, in order to test its effectiveness (and also to facilitate the work of the core team). I am available to make any additional adjustments to it if necessary.

### CODE REVIEW

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (NataliaTepluhina) [inline src/.vuepress/components/common/codepen-snippet.vue:37]: **nitpick**: should we make `title` and `slug` required? This way it would be easier to see that we didn't miss important ones.

**Review by NataliaTepluhina — APPROVED** [review]
@ErickPetru I like the idea! I've left one nitpick to codepen component but overall it looks good to me 👍🏻

File: src/guide/introduction.md
Comment (skirtles-code) [inline src/guide/introduction.md:183]: I think this needs `tab="js,result"`.

This particular example encourages the reader to edit the code.

File: src/guide/transitions-overview.md
Comment (skirtles-code) [inline src/guide/transitions-overview.md:177]: There seem to be some changes to this example. Were these intentional?

File: src/examples/commits.md
Comment (skirtles-code) [inline src/examples/commits.md:5]: Does this need `tab="js,result"`?

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (skirtles-code) [inline src/.vuepress/components/common/codepen-snippet.vue:6]: Some of the current examples are not editable. Specifically, those in the Examples section.

Is changing this intentional?

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (skirtles-code) [inline src/.vuepress/components/common/codepen-snippet.vue:4]: Some of the current examples use a different value here.

I'm not sure what it does but is the change intentional?

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (skirtles-code) [inline src/.vuepress/components/common/codepen-snippet.vue:5]: A lot of the current examples don't have this.

Again, I'm not sure what it does but I was wondering whether the change was intentional?

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (skirtles-code) [inline src/.vuepress/components/common/codepen-snippet.vue:14]: Perhaps replace `name ? name : user` with `name || user`?

File: src/guide/transitions-overview.md
Comment (skirtles-code) [inline src/guide/transitions-overview.md:73]: Should the word 'Lists' be included in this title?

File: src/guide/transitions-enterleave.md
Comment (skirtles-code) [inline src/guide/transitions-enterleave.md:137]: I think there's been a copy/paste error here with the `title` and `slug`.

File: src/guide/a11y-basics.md
Comment (skirtles-code) [inline src/guide/a11y-basics.md:59]: There are a few examples that used to mention 'Maria' that don't seem to be doing that anymore. Is this intentional?

**Review by skirtles-code — CHANGES_REQUESTED** [review]
A few suggestions from me.

**Review by bencodezen — CHANGES_REQUESTED** [review]
Thanks for the work on this @ErickPetru! Your refactoring has shown us that there's some clean up that needs to do in terms of standardizing some things so it's easier to turn into a component. I think we should probably hold on this until those changes are made so you can update the PR.

File: src/guide/introduction.md
Comment (ErickPetru) [reply src/guide/introduction.md]: Yes, the original has `data-default-tab="js,result"`. I just forgot to add with the `tab` prop.

File: src/guide/transitions-overview.md
Comment (ErickPetru) [reply src/guide/transitions-overview.md]: Thanks, by mistake I forgot to add the custom author as props in this case.

File: src/examples/commits.md
Comment (ErickPetru) [reply src/examples/commits.md]: Yes, added.

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (ErickPetru) [reply src/.vuepress/components/common/codepen-snippet.vue]: Initially I thought we could define a standard to it, but after more thinking I realized that each section can have its own reasons to allow or not inline editable examples. So, my intention is not to change any current behavior, just improve future maintenance with the component. My recent commits added this behavior as props too.

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (ErickPetru) [reply src/.vuepress/components/common/codepen-snippet.vue]: I really think its a problem, the standardized theme should be used along all the docs. But I don't want to change behavior with this PR, so I've created a new `theme` prop too. And it'll be easy to solve the different theme in some files in the future, if the team wants.

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (ErickPetru) [reply src/.vuepress/components/common/codepen-snippet.vue]: Yes, the same case as I said about `editable`. Initially I thought `preview` should be always true. But my intention here isn't changing already established behavior, so I updated all files that hadn't `data-preview` putting the prop `:preview="false"`.

File: src/.vuepress/components/common/codepen-snippet.vue
Comment (ErickPetru) [reply src/.vuepress/components/common/codepen-snippet.vue]: Nice catch, tanks.

File: src/guide/transitions-enterleave.md
Comment (ErickPetru) [reply src/guide/transitions-enterleave.md]: For sure, corrected.

File: src/guide/a11y-basics.md
Comment (ErickPetru) [reply src/guide/a11y-basics.md:59]: No, just forgot the `name` prop there and I already updated in the recent commits. But in the future I think the core team could migrate everything to the **team/Vue** Codepen account.

**Comment by ErickPetru** [comment]
Hi @bencodezen, my intention with this PR was not to change current behavior.

I really forgot some possible adaptations at the first pass, but after @skirtles-code review, I've already sent more commits with everything (I hope) in the same way it was before.

I believe it's secure to merge after the next review phase. Indeed, even future standardization on the embeds will be a lot easier with the new component. Also, feel free to ask for more changes if you want.

File: src/guide/transitions-overview.md
Comment (ErickPetru) [reply src/guide/transitions-overview.md]: No, another mistake.

**Comment by skirtles-code** [comment]
@ErickPetru Currently there are conflicts with `master`.

Given what @bencodezen said, I'm unclear whether more conflicts are coming or if this is now good to proceed.

I'm happy to do another review but I'd rather wait for confirmation that there aren't more conflicts coming first.

**Comment by ErickPetru** [comment]
@skirtles-code and @bencodezen I saw the recent changes on `examples/modal.md` and `examples/grid-component.md` adding the updated Codepen URLs. So I just sent these two updated using the CodepenSnippet component.

File: src/guide/transitions-state.md
Comment (skirtles-code) [inline src/guide/transitions-state.md:56]: The original example had `data-height="500"` but `style="height: 493px;"`.

From checking the deployed version of this branch it seems that `:height="500"` would be a better match for the live docs.

File: src/cookbook/editable-svg-icons.md
Comment (skirtles-code) [inline src/cookbook/editable-svg-icons.md:146]: The `theme` for this example was previously set to `0`.

File: src/guide/transitions-state.md
Comment (skirtles-code) [inline src/guide/transitions-state.md:145]: This was previously using the `light` theme.

**Review by skirtles-code — CHANGES_REQUESTED** [review]
Some more suggestions from me.

I've not made any subjective decisions about whether things matter, I've just reported anything that seems to have changed from what was there previously.

I agree that this component should make it easier to remove the current inconsistencies.

**Comment by ErickPetru** [comment]
Sorry for the delay, hard job weeks here currently.
I've just updated with the corrections after the second @skirtles-code review.

**Review by skirtles-code — CHANGES_REQUESTED** [review]
There's a new CodePen example at the bottom of `transitions-enterleave.md`.

Usually I'd suggest migrating it post-merge but in this case it seems that the component can't co-exist with non-component CodePens. That new example gets a thick, black border around it as a result.

I did a bit of debugging and it seems to be a consequence of how SSR and the CodePen script interact with each other. So long as everything is using the component it should be fine.

As SSR is involved you can only see the problem by going directly to the right page:

https://deploy-preview-548--vue-docs-next-preview.netlify.app/guide/transitions-enterleave.html#transitioning-between-components

If you navigate there via another page it all behaves.

**Comment by ErickPetru** [comment]
Incredible was your catch about SSR with Codepen embeds without the new component, @skirtles-code!

So as you suggested, I just updated `guide/transitions-enterleave.md` changing the last embed to also use the component.

Then I just realized that the project could have direct `<iframe>` embeds to, and I was right. So I changed to use the component also in the files `guide/optimizations.md` and `guide/reactivity.md`. While the way it works becomes a little different (from an direct `<iframe>` to a `<p>` converted to an `<iframe>` by https://static.codepen.io/assets/embed/ei.js), in my tests it offers the same experience to the end user and finally standardize all the guide to use the `CodepenSnippet` component.

Please take a look and talk to me if something more is needed.

**Review by bencodezen — APPROVED** [review]
Thanks for your hard work on this @ErickPetru!

**Pull request merged**
